### PR TITLE
Fix shell workaround

### PR DESCRIPTION
**Pull request by ybiquitous**
## Summary

- `child_process.spawn()` of Node.js supports the `shell` option to reduce platform-dependent code.
  See https://nodejs.org/api/child_process.html#child_processspawncommand-args-options
- Test other platforms than Linux on CI

## My problem

I encounter the following error on my macOS machine, and so this extension doesn't work. I believe this PR can fix the problem.

```
[vscode] Try to start TypeProf for IDE
[vscode] stderr: (eval):1: can't change option: zle
[vscode] stderr: bundler: failed to load command: typeprof (/Users/masafumi.koba/.rbenv/versions/3.3.0-preview1/bin/typeprof)
[vscode] stderr: /Users/masafumi.koba/.rbenv/versions/3.3.0-preview1/lib/ruby/3.3.0+0/bundler/rubygems_integration.rb:308:in `block in replace_bin_path': can't find executable typeprof for gem typeprof. typeprof is not currently included in the bundle, perhaps you meant to add it to your Gemfile? (Gem::Exception)
```

### CODE REVIEW

File: .github/workflows/vscode.yml [hunk 1]
```diff
@@ -14,30 +14,39 @@ on:
...
+        bundler: latest
     - name: Install Node.js
       uses: actions/setup-node@v3
       with:
         node-version: 16.x
-    - run: gem install bundler
-    - run: npm install
+        cache: npm
+    - name: Install npm packages
+      run: npm ci
```
Comment (ybiquitous) [inline .github/workflows/vscode.yml:37]: [note] [`npm ci`](https://docs.npmjs.com/cli/v9/commands/npm-ci) respects `package-lock.json`. This should be preferable on CI.

File: .github/workflows/vscode.yml [hunk 2]
```diff
@@ -14,30 +14,39 @@ on:
...
       run: npm run lint:prettier
-    - name: Test
+    - name: Test with Xvfb
       run: xvfb-run -a npm test
+      if: ${{ runner.os == 'Linux' }}
+    - name: Test
+      run: npm test
+      if: ${{ runner.os != 'Linux' }}
+      env:
+        RUBY_BIN_PATH: ${{ runner.os == 'macOS' && format('{0}/{1}', steps.setup-ruby.outputs.ruby-prefix, 'bin') || '' }}
```
Comment (ybiquitous) [inline .github/workflows/vscode.yml:49]: [note] Workaround for https://github.com/microsoft/vscode-test/issues/20

**Comment by ono-max** [comment]
@ybiquitous 

Thank you for your work. If you're ok, can you divide PR into the shell part and test part?

**Comment by ybiquitous** [comment]
Of course, I'd gladly split the pull request if CI had no problems...

**Comment by ono-max** [comment]
From my investigation, it seems to be that the reason for failure in CI is that the path of `bundler` is wrong.

Expected Path: `/Users/runner/hostedtoolcache/Ruby/3.1.0/x64/bin/bundler`

Actual Path: ` /usr/local/lib/ruby/gems/3.0.0/gems/bundler-2.4.12/lib/bundler/`

**Comment by ybiquitous** [comment]
@ono-max Thanks for the investigation. It seems a `PATH` problem... 🤔 

`process.env.PATH` changed [here](https://github.com/ruby/vscode-typeprof/pull/103/files#diff-a2dc793ec9bb9e46679a7dbbef11723eeb9416a48b7b0dfb753e5760cdb5b923R12) may not correctly be passed to `child_process.spawn()`:

https://github.com/ruby/vscode-typeprof/blob/3a0ac8eb005ee54d2f0a8ba3c517cfd91f406ee8/src/extension.ts#L115

I created this PR, which passes the `shell` option to `child_process.spawn()`, to avoid this `PATH` problem.
See also <https://github.com/ruby/vscode-typeprof/pull/103/files#diff-04bba6a35cad1c794cbbe677678a51de13441b7a6ee8592b7b50be1f05c6f626R109>.

File: src/test/suite/index.ts [hunk 3]
```diff
@@ -3,6 +3,17 @@ import * as Mocha from 'mocha';
...
 export function run(): Promise<void> {
+    if (process.env.PATH && process.env.RUBY_BIN_PATH) {
+        console.log(
+            "Prefer a newly-installed Ruby version over the system's default one." +
+                ' This is a workaround for https://github.com/microsoft/vscode-test/issues/20',
+        );
+        const previous = process.env.PATH;
+        process.env.PATH = [process.env.RUBY_BIN_PATH, ...previous.split(path.delimiter)].join(path.delimiter);
+        console.log(`Previous PATH: ${previous}`);
+        console.log(` Current PATH: ${process.env.PATH}`);
```
Comment (ono-max) [inline src/test/suite/index.ts:14]: nit
```suggestion
        console.log(`Current PATH: ${process.env.PATH}`);
```

**Comment by ono-max** [comment]
@ybiquitous 

I see. Thank you for your explanation. I'll merge this PR, and you don't have to split PR since all CI passed.

By the way, could you tell me why this change solves the `PATH` problem?

File: src/test/suite/index.ts (hunk 3 above)
Comment (mame) [reply src/test/suite/index.ts:14]: I think this is matching the columns of the colon. I am fine with the original code.

**Comment by mame** [comment]
Does the `shell` option for `child_process.spawn` read .bashrc/.zshrc?

Many users write PATH configuration in .bashrc/.zshrc, like `export PATH=/path/to/my/ruby:$PATH` or `eval "$(~/.rbenv/bin/rbenv init - bash)"` or something. It is important for vscode-typeprof to use the configuration. This is why vscode-typeprof invokes the shell in interactive mode, i.e., `bash -l -c` or `zsh -i -l -c`.

If the `shell` option for `child_process.spawn` still has the same effect, it's OK. If not, then we need to consider.

**Comment by ybiquitous** [comment]
> By the way, could you tell me why this change solves the `PATH` problem?

To be honest, I don't understand why the `shell` option can resolve the `PATH` problem... 😓 

> Does the `shell` option for `child_process.spawn` read .bashrc/.zshrc?

Umm, I guess reading .bashrc/.zshrc may cause a side-effect like `can't change option: zle`. 🤔 Any thoughts?

**Comment by ybiquitous** [comment]
@mame Oh, sorry for leaving this problem. 🙇🏼 

I found out `(eval):1: can't change option: zle` is not a real problem. The following output is emitted on my local machine.

```
[vscode] Try to start TypeProf for IDE
[vscode] stderr: (eval):1: can't change option: zle
[vscode] typeprof version: ]1; typeprof 0.21.7
[vscode] typeprof --version showed unknown message
[vscode] Ruby TypeProf is not configured
```

The real problem here is the `typeprof version: ...` line. It includes escape sequences somehow... 🤷🏼 See also the screenshot below:

Screenshot 2023-05-30 at 11 34 25

The following patch works on my machine:

```diff
--- a/src/extension.ts
+++ b/src/extension.ts
@@ -152,7 +152,7 @@ function getTypeProfVersion(
         if (code === 0) {
             const str = output.trim();
             log(`typeprof version: ${str}`);
-            const version = /^typeprof (\d+.\d+.\d+)$/.exec(str);
+            const version = /typeprof (\d+.\d+.\d+)$/.exec(str);
             if (version && version.length === 2) {
                 if (compareVersions(version[1], '0.20.0') >= 0) {
                     callback(null, version[1]);
```

---

FYI. I find that `(eval):1: can't change option: zle` is emitted by [`fzf`](https://github.com/junegunn/fzf). This has no actual effects, although it shows only on the VSCode TypeProf extension. I'm unsure of this cause, too.

**Comment by ybiquitous** [comment]
FYI. You can see my `.zshrc` here (using oh-my-zsh): https://github.com/ybiquitous/dotfiles/blob/8a86ab9fe431de85b53dd84142df28bbe8c08f41/HOME/.zshrc#L85

Disabling the `fzf` plugin for oh-my-zsh like this can disappear the error message `(eval):1: can't change option: zle`:

```diff
 plugins=(
   bundler
   direnv
-  fzf
+#  fzf
   history
   history-substring-search
   npm
```

**Comment by mame** [comment]
Thanks. Then, let's ignore the `can't change option: zle`.

However, I guess that the problem is not solved yet.

vscode-typeprof first invokes `bundle exec typeprof --version` to check the version, and if it is OK, then invokes `bundle exec typeprof --lsp` to start communication.  If the output of `typeprof --version` contains escape sequences, probably the output of `typeprof --lsp` also does. vscode-typeprof tries to `JSON.parse` the output of `typeprof --lsp`, but if it fails, the error is just ignored by `try { ... } catch (err) {}`. (My bad. This should be fixed.)

Perhaps someone in .zshrc is emitting the `<ESC>]1; <BEL>` escape sequence. I wonder if vscode-typeprof needs to skip such escape sequences.

**Comment by ybiquitous** [comment]
As you imagined, I confirmed the output of `typeprof --lsp` included escape sequences. But, it still works. See below:

Output on VSCode:

```
[vscode] Try to start TypeProf for IDE
[vscode] stderr: (eval):1: can't change option: zle
[vscode] typeprof version: ]1; typeprof 0.21.7
[vscode] Starting Ruby TypeProf (0.21.7)...
[Info  - 1:04:09 PM] (eval):1: can't change option: zle
[vscode] Received data: ]1; 
[vscode] SyntaxError: Unexpected token '', "]1; " is not valid JSON
[vscode] Received data: {"host":"::1","port":52285,"pid":11105}
[vscode] Received JSON: {"host":"::1","port":52285,"pid":11105}
[Info  - 1:04:11 PM] TypeProf for IDE is started successfully
[vscode] Ruby TypeProf is running
```

Screenshot:

image

Note that I changed the code temporarily for debugging:

```diff
--- a/src/extension.ts
+++ b/src/extension.ts
@@ -123,6 +123,11 @@ function executeTypeProf(folder: vscode.WorkspaceFolder, arg: String): child_pro
     return typeprof;
 }
 
+const log = (msg: string) => {
+    outputChannel.appendLine('[vscode] ' + msg);
+    console.info(msg);
+};
+
 function getTypeProfVersion(
     folder: vscode.WorkspaceFolder,
     callback: (err: Error | null, version: string) => void,
@@ -130,11 +135,6 @@ function getTypeProfVersion(
     const typeprof = executeTypeProf(folder, '--version');
     let output = '';
 
-    const log = (msg: string) => {
-        outputChannel.appendLine('[vscode] ' + msg);
-        console.info(msg);
-    };
-
     typeprof.stdout?.on('data', (out) => {
         output += out;
     });
@@ -152,7 +152,7 @@ function getTypeProfVersion(
         if (code === 0) {
             const str = output.trim();
             log(`typeprof version: ${str}`);
-            const version = /^typeprof (\d+.\d+.\d+)$/.exec(str);
+            const version = /typeprof (\d+.\d+.\d+)$/.exec(str);
             if (version && version.length === 2) {
                 if (compareVersions(version[1], '0.20.0') >= 0) {
                     callback(null, version[1]);
@@ -187,12 +187,16 @@ function getTypeProfStream(
 
         let buffer = '';
         typeprof.stdout.on('data', (data) => {
+            log(`Received data: ${data}`);
             buffer += data;
             try {
                 const json = JSON.parse(data);
                 json['stop'] = () => typeprof.kill('SIGINT');
+                log(`Received JSON: ${JSON.stringify(json)}`);
                 resolve(json);
-            } catch (err) {}
+            } catch (err) {
+                log(`${err}`);
+            }
         });
 
         let err = '';
```

File: src/extension.ts [hunk 4]
```diff
@@ -83,44 +84,29 @@ function showErrorStatusBar() {
...
-            args.push('-i');
-        }
-        args.push('-l', '-c', cmd);
-        typeprof = child_process.spawn(shell, args, { cwd });
-    } else if (process.platform === 'win32') {
-        typeprof = child_process.spawn(process.env.SYSTEMROOT + '\\System32\\cmd.exe', ['/c', cmd], { cwd });
-    } else {
-        const cmds = cmd.split(' ');
-        typeprof = child_process.spawn(cmds[0], cmds.slice(1), { cwd });
-    }
```
Comment (ybiquitous) [inline src/extension.ts:121]: [note] I try looking into the `shell` option more (Node.js internal). See the code in Node.js:

```js
  if (options.shell) {
    validateArgumentNullCheck(options.shell, 'options.shell');
    const command = ArrayPrototypeJoin([file, ...args], ' ');
    // Set the shell, switches, and commands.
    if (process.platform === 'win32') {
      if (typeof options.shell === 'string')
        file = options.shell;
      else
        file = process.env.comspec || 'cmd.exe';
      // '/d /s /c' is used only for cmd.exe.
      if (RegExpPrototypeExec(/^(?:.*\\)?cmd(?:\.exe)?$/i, file) !== null) {
        args = ['/d', '/s', '/c', `"${command}"`];
        windowsVerbatimArguments = true;
      } else {
        args = ['-c', command];
      }
    } else {
      if (typeof options.shell === 'string')
        file = options.shell;
      else if (process.platform === 'android')
        file = '/system/bin/sh';
      else
        file = '/bin/sh';
      args = ['-c', command];
    }
  }
```

https://github.com/nodejs/node/blob/49412c9632dc697ff52a3f731c578afc26c4654b/lib/child_process.js#L617-L642

This Node.js code might be helpful if we would not use the `shell` option, especially for Windows or Android.

---

Now I understand the need to work around shells (e.g. `zsh -i -l -c "..."`), so I'll close this PR later (I'll keep it open for a while for other people who encounter similar problems).

**Pull request closed without merging**